easy_transformer/hook_points.py

```diff
@@ -1,149 +1,216 @@
                 return hook(module_output, hook=self)
 
             handle = self.register_forward_hook(full_hook)
-            self.fwd_hooks.append(handle)
-        elif dir == "bwd":
-            # For a backwards hook, module_output is a tuple of (grad,) - I don't know why.
+            self.forward_hooks.append(handle)
+        # TODO find a tool that complains when i try to test two different-typed things for equality
+
+        elif direction == HookDirection.BACKWARD:
+
             def full_hook(module, module_input, module_output):
-                return hook(module_output[0], hook=self)
+                # For a backwards hook, module_output is a tuple of (grad,) - I don't know why.
+                grad = module_output[0]
+                return hook(grad, hook=self)
 
             handle = self.register_full_backward_hook(full_hook)
-            self.bwd_hooks.append(handle)
+            self.backward_hooks.append(handle)
         else:
-            raise ValueError(f"Invalid direction {dir}")
-
-    def remove_hooks(self, dir="fwd"):
-        if (dir == "fwd") or (dir == "both"):
-            for hook in self.fwd_hooks:
+            # TODO for neel nanda- do we want to allow [HookDirection.BOTH]?
+            raise ValueError(f"Invalid direction {direction}")
+
+    def remove_hooks(self, direction: HookDirection = HookDirection.FORWARD):
+        if not isinstance(direction, HookDirection):
+            raise ValueError(f"Invalid direction {direction}")
+        if (direction == HookDirection.FORWARD) or (direction == HookDirection.BOTH):
+            for hook in self.forward_hooks:
                 hook.remove()
-            self.fwd_hooks = []
-        if (dir == "bwd") or (dir == "both"):
-            for hook in self.bwd_hooks:
+            self.forward_hooks = []
+        if (direction == HookDirection.BACKWARD) or (direction == HookDirection.BOTH):
+            for hook in self.backward_hooks:
                 hook.remove()
-            self.bwd_hooks = []
-        if dir not in ["fwd", "bwd", "both"]:
-            raise ValueError(f"Invalid direction {dir}")
+            self.backward_hooks = []
 
     def clear_context(self):
-        del self.ctx
-        self.ctx = {}
+        del self.context
+        self.context = {}
 
     def forward(self, x):
+        """Identity function"""
         return x
 
     def layer(self):
-        # Returns the layer index if the name has the form 'blocks.{layer}.{...}'
-        # Helper function that's mainly useful on EasyTransformer
-        # If it doesn't have this form, raises an error -
-        split_name = self.name.split(".")
-        return int(split_name[1])
+        """
+        Returns the layer index if the name has the form 'blocks.{layer}.{...}'
+        Helper function that's mainly useful on EasyTransformer
+        If it doesn't have this form, raises an error -
+        """
+        _, layer, _ = self.name.split(".")
```

Comment:
This won't work, the list may be three or four elements long (also this should probably return None for hooks without a layer, rather than error lol)

Reply:
6f67168

Reply:
how do you like this style? 23fbe80812c674d9ba522e3a707985f2e43015fa

Reply:
(i will finish the rest if you approve!)